### PR TITLE
Replace ParmEd with Interchange in Amber/GROMACS example

### PR DESCRIPTION
**Pull request by mattwthompson**
- [x] Re-reun once Interchange v0.1.2 is online
- [ ] Tag issue being addressed
- [x] Add [tests](https://github.com/openforcefield/openff-toolkit/tree/master/openff/toolkit/tests)
- [ ] Update docstrings/[documentation](https://github.com/openforcefield/openff-toolkit/tree/master/docs), if applicable
- [x] [Lint](https://open-forcefield-toolkit.readthedocs.io/en/latest/developing.html#style-guide) codebase
- [x] Update [changelog](https://github.com/openforcefield/openff-toolkit/blob/master/docs/releasehistory.rst)

### CODE REVIEW

**Comment by mattwthompson** [comment]
I can't reproduce the test failures locally, even with downgrading to Python 3.7 and in a similar-ish-looking environment: https://gist.github.com/mattwthompson/eaef9191b31ce891816acfdb6e93f50e

**Comment by j-wags** [comment]
Yeah, QCArchive seems to be down/unreachable. MolSSI is aware and they're working on resolving.

**Comment by mattwthompson** [comment]
I'm getting errors from Sander in these tests, though:


```

_ examples/using_smirnoff_in_amber_or_gromacs/export_with_interchange.ipynb::Cell 9 _
Notebook cell execution failed
Cell 9: Cell execution caused an exception

Input:
get_amber_energies(interchange).energies

Traceback:

---------------------------------------------------------------------------
SanderError                               Traceback (most recent call last)
/tmp/ipykernel_4254/100924973.py in <module>
----> 1 get_amber_energies(interchange).energies

/usr/share/miniconda3/envs/test/lib/python3.7/site-packages/openff/interchange/drivers/amber.py in get_amber_energies(off_sys, writer, electrostatics)
     69                 inpcrd_file="out.inpcrd",
     70                 in_file=in_file,
---> 71                 electrostatics=electrostatics,
     72             )
     73             return report

/usr/share/miniconda3/envs/test/lib/python3.7/site-packages/openff/interchange/drivers/amber.py in _run_sander(inpcrd_file, prmtop_file, in_file, electrostatics)
    116 
    117     if sander.returncode:
--> 118         raise SanderError
    119 
    120     energies, _ = _group_energy_terms("mdinfo")

SanderError: 

_ examples/using_smirnoff_in_amber_or_gromacs/export_with_interchange.ipynb::Cell 10 _
Notebook cell execution failed
Cell 10: Cell execution caused an exception

Input:
get_summary_data(interchange)

Traceback:

---------------------------------------------------------------------------
SanderError                               Traceback (most recent call last)
/tmp/ipykernel_4254/736456265.py in <module>
----> 1 get_summary_data(interchange)

/usr/share/miniconda3/envs/test/lib/python3.7/site-packages/openff/utilities/utilities.py in wrapper(*args, **kwargs)
     71                 raise e
     72 
---> 73             return function(*args, **kwargs)
     74 
     75         return wrapper

/usr/share/miniconda3/envs/test/lib/python3.7/site-packages/openff/interchange/drivers/all.py in get_summary_data(interchange)
     39     kj_mol = unit.kilojoule / unit.mol
     40 
---> 41     energies = get_all_energies(interchange)
     42 
     43     for k, v in energies.items():

/usr/share/miniconda3/envs/test/lib/python3.7/site-packages/openff/interchange/drivers/all.py in get_all_energies(interchange)
     27         "GROMACS": get_gromacs_energies(interchange),
     28         "LAMMPS": get_lammps_energies(interchange),
---> 29         "Amber": get_amber_energies(interchange, writer="internal"),
     30     }
     31 

/usr/share/miniconda3/envs/test/lib/python3.7/site-packages/openff/interchange/drivers/amber.py in get_amber_energies(off_sys, writer, electrostatics)
     69                 inpcrd_file="out.inpcrd",
     70                 in_file=in_file,
---> 71                 electrostatics=electrostatics,
     72             )
     73             return report

/usr/share/miniconda3/envs/test/lib/python3.7/site-packages/openff/interchange/drivers/amber.py in _run_sander(inpcrd_file, prmtop_file, in_file, electrostatics)
    116 
    117     if sander.returncode:
--> 118         raise SanderError
    119 
    120     energies, _ = _group_energy_terms("mdinfo")

SanderError: 
```

which do not occur locally.

**Comment by j-wags** [comment]
Ahh, right, those should be independent from the QCA issues. Sorry - my brain is booting up slowly today. I'm looking into what's happening here.

**Comment by j-wags** [comment]
Hm, one commit ago it was failing to find `sander` and `lmp_serial`. But as of the last commit (where I added the LAMMPS dep to the right yaml) it's just failing to find `sander`. So it IS picking up the LAMMPS dependency, but maybe something about the GROMACS/LAMMPS install is breaking AmberTools? 

I'm also noticing that the new notebook PASSES on macos with RDKit only. Here's the breakdown:
* macos, RDKit = Pass
* macos, OpenEye = Fail w/ sander error
* linux, RDKit = Fail w/ sander error
* linux, OpenEye = Fail w/ sander error

**Comment by mattwthompson** [comment]
No worries! I haven't implemented great error handling yet, since sometimes engines yell at me via STDOUT, sometimes via STDERR, and sometimes they're quiet and hide errors in a file.

If you or somebody else has time, It would be useful to see if this can be reproduced locally; if you can't, then I'll have to work on error handling and come back to this later.

**Comment by mattwthompson** [comment]
The code to find `sander` is naive: https://github.com/openforcefield/openff-interchange/blob/v0.1.2/openff/interchange/drivers/amber.py#L103

My personal experience has always been that having AmberTools installed also provides Sander. I've never had issues with LAMMPS or GROMACS installs conflicting


... oh, now I see that the non-RDKit builds forcefully remove AmberTools. 🤦  That'll explain at least two of the three errors in the matrix.

**Comment by j-wags** [comment]
> Does our distribution of examples as a conda package mean we can only have conda-forge-installable packages in our examples?

Good catch. Not quite. We already have a GROMACS section in the "appendix" of the `toolkit_showcase` example, which just won't run out of the box without the additional dep. So it'd probably be best to separate the GROMACS-dependent cell of the interchange notebook into an appendix, with instructions on how to install it.

**Comment by mattwthompson** [comment]
Oof ...

```


=================================== FAILURES ===================================
_ examples/using_smirnoff_in_amber_or_gromacs/export_with_interchange.ipynb::Cell 8 _
Notebook cell execution failed
Cell 8: Cell execution caused an exception

Input:
amber_energies = get_amber_energies(interchange)
amber_energies.energies

Traceback:

---------------------------------------------------------------------------
SanderError                               Traceback (most recent call last)
/tmp/ipykernel_5053/3004181806.py in <module>
----> 1 amber_energies = get_amber_energies(interchange)
      2 amber_energies.energies

/usr/share/miniconda3/envs/test/lib/python3.7/site-packages/openff/interchange/drivers/amber.py in get_amber_energies(off_sys, writer, electrostatics)
     74                 inpcrd_file="out.inpcrd",
     75                 in_file=in_file,
---> 76                 electrostatics=electrostatics,
     77             )
     78             return report

/usr/share/miniconda3/envs/test/lib/python3.7/site-packages/openff/interchange/drivers/amber.py in _run_sander(inpcrd_file, prmtop_file, in_file, electrostatics)
    127 
    128     if sander.returncode:
--> 129         raise SanderError(err)
    130 
    131     energies, _ = _group_energy_terms("mdinfo")

SanderError: double free or corruption (out)

Program received signal SIGABRT: Process abort signal.

Backtrace for this error:
#0  0x7f2ba75e220f in ???
#1  0x7f2ba75e218b in ???
#2  0x7f2ba75c1858 in ???
#3  0x7f2ba762c3ed in ???
#4  0x7f2ba763447b in ???
#5  0x7f2ba763611f in ???
#6  0x55adbb8045a2 in ???
#7  0x55adbb800984 in ???
#8  0x55adbb8009fc in ???
#9  0x7f2ba75c30b2 in ???
#10  0x55adbb636098 in ???
Aborted (core dumped)


=========================== short test summary info ============================
FAILED examples/using_smirnoff_in_amber_or_gromacs/export_with_interchange.ipynb::Cell 8
============ 1 failed, 168 passed, 3 skipped in 1145.69s (0:19:05) =============
Error: Process completed with exit code 1.

**Comment by mattwthompson** [comment]
I'm running out of resources tracking down this behavior; I've gone so far as printing out the files themselves and nothing seems out of order. They pass the eye check. I can copy+paste them into a new folder and run them just fine, both directly and via `_run_sander`. Running tests on the notebook gives me only passing tests. This is all using the same branch that this PR sees. Downgrading to Python 3.7 does not seem to change anything.

**Comment by mattwthompson** [comment]
Nothing stands out to me in the AmberTools changelog: https://ambermd.org/ATPatches.php

**Comment by j-wags** [comment]
Here are some possibilities I can think of:
* a filesystem/file buffer issue? Like, maybe the amber inputs aren't fully written before sander tries to load them, and the rdkit/ubuntu configuration already has the file system under heavy load?
* a newly-discovered conflict with rdkit, gromacs, or lammps? I went looking through previous CI runs on this PR but couldn't find one that got far enough to trigger this sander error
* _something something_ trying pinning to an older ambertools?

File: examples/environment.yaml [hunk 1]
```diff
@@ -17,3 +17,6 @@ dependencies:
   - numpy
   - openmm
   - openff-forcefields >=2.0.0
+  - openff-interchange >=0.1.2
+  - gromacs
+  - lammps
```
Comment (j-wags) [inline examples/environment.yaml:22]:
```suggestion
  - lammps
  - ambertools =21.3
```

File: .github/workflows/examples.yml [hunk 2]
```diff
@@ -61,7 +61,7 @@ jobs:
           auto-activate-base: false
           miniforge-version: latest
           miniforge-variant: Mambaforge
-          use-mamba: true
+          use-mamba: false
```
Comment (j-wags) [inline .github/workflows/examples.yml:64]:
```suggestion
          use-mamba: true
```

**Comment by mattwthompson** [comment]
Lily suggested running the notebook as a script - with a couple steps (`jupyter nbconvert --to script export_with_interchange.ipynb` and pointing to the PDB file in a safer way) I got it online. Still working locally

```
(openff-dev) [openforcefield] pytest -q openff/toolkit/tests/test_examples.py                   drop-parmed-examples  ✱
...                                                                                                              [100%]
3 passed in 31.94s
(openff-dev) [openforcefield] python -m pytest -q examples/using_smirnoff_in_amber_or_gromacs/export_with_interchange.ipynb --nbval-lax -p no:randomly --current-env
.............                                                                                                    [100%]
13 passed in 13.66s
```
If it fails in CI as a script, there is something wrong with the actual code and I'll need to revisit how I'm using subprocesses. If not, there is something funky happening with the notebook and/or how it's called from `pytest`.

**Comment by mattwthompson** [comment]
Thanks again for helping with this. The way I call other programs is relatively naive, but it's pretty boilerplate and hasn't given me much trouble in the past:

https://github.com/openforcefield/openff-interchange/blob/33a5e3a0b3009c10beab1884ff6ee58aca3e6453/openff/interchange/drivers/amber.py#L170-L178

The system is small (two molecules) so hardware limits shouldn't be the issue. Memory leaking at some point in the process (an executable starting tests, which calls something else to run the notebook, which then further spawns a new subprocess) sounds totally plausible, if hard to hunt down. I'm updating an old Linux box to see if I can reproduce it locally. Some obscure OS difference would explain why it works for me locally (normally using macOS) but fails on CI.

**Comment by mattwthompson** [comment]
I can reproduce this on an machine running Linux, and one fix is to remove the box information from the coordinate file. I'll need to gather information on if people want box information in both the coordinate and topology file.

File: .github/workflows/examples.yml [hunk 3]
```diff
@@ -101,6 +101,8 @@ jobs:
       - name: Install package
         shell: bash -l {0}
         run: |
+          pip install git+https://github.com/openforcefield/openff-interchange.git@skip-drivers
+
```
Comment (mattwthompson) [inline .github/workflows/examples.yml:105]: To be explicit: I will include this branch into a release before aiming to have this PR merged

**Comment by mattwthompson** [comment]
The issue, or at least what I fixed locally to get it working, is this: https://github.com/openforcefield/openff-interchange/pull/332/commits/6e4fd122213eb089f81748ca8bf84591c7772c07

File: .github/workflows/examples.yml (hunk 3 above)
Comment (mattwthompson) [reply .github/workflows/examples.yml:105]: This is now included in v0.1.3; CI will need to be restarted next week to get it to pass (the package will not be online until after hours)

**Comment by mattwthompson** [comment]
Failures in last run are codecov.

**Pull request merged**